File: GeoFencer/Geofence.m
#import "Geofence.h"

@implementation Geofence

- (instancetype)initWithPoints:(NSArray<MKPointAnnotation *> *)points {
    self = [super init];
    if (!self) return nil;

    _points = points;

    return self;
}

- (MKCircle *)circle {
    return [self.class circleFromPoints:self.points];
}

+ (MKCircle *)circleFromPoints:(NSArray<MKPointAnnotation *> *)points  {
    CGFloat minX = CGFLOAT_MAX,
        maxX = CGFLOAT_MIN,
        minY = CGFLOAT_MAX,
        maxY = CGFLOAT_MIN;

    for (MKPointAnnotation *point in points) {
        MKMapPoint p = MKMapPointForCoordinate(point.coordinate);
        minX = fmin(minX, p.x);
        minY = fmin(minY, p.y);
        maxX = fmin(maxX, p.x);
        maxY = fmin(maxY, p.y);
    }

    MKMapRect rect = MKMapRectMake(minX,
                                 minY,
                                 maxX - minX,
                                 maxY - minY);
    return [MKCircle circleWithMapRect:rect];
}

@end

File: GeoFencer/Geofence.swift
import Foundation
import MapKit

class Geofence {
    let points:[CLLocationCoordinate2D]
    let circle: MKCircle

    init (points:[CLLocationCoordinate2D]) {
        self.points = points

        // TODO: This assumes only 2 points
        circle = Geofence.circleFromPoints(points.first!, points.last!)
    }

    class func circleFromPoints(first:CLLocationCoordinate2D, _ second:CLLocationCoordinate2D) -> MKCircle {
        let p1 = MKMapPointForCoordinate(first)
        let p2 = MKMapPointForCoordinate(second)

        let mapRect = MKMapRectMake(fmin(p1.x,p2.x),
                                    fmin(p1.y,p2.y),
                                    fabs(p1.x-p2.x),
                                    fabs(p1.y-p2.y))
        return MKCircle(mapRect:mapRect)
    }
}
